### PR TITLE
Skip adding null during Insert if the column has default value

### PR DESCRIPTION
**Pull request by Praveen2112**
Fises #1185

### CODE REVIEW

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java [hunk 1]
```diff
@@ -207,6 +207,20 @@ public void testInsertIntoNotNullColumn()
         assertUpdate("DROP TABLE test_insert_not_null");
     }
 
+    @Test
+    public void testInsertForDefaultColumn()
+            throws Exception
+    {
+        execute("CREATE TABLE tpch.default_table (a bigint, b bigint default 40)");
+        assertUpdate("INSERT INTO default_table (a) VALUES (1)", 1);
+        assertUpdate("INSERT INTO default_table VALUES (2, 3)", 1);
```
Comment (kokosing) [inline presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java]: test with insert `(4, null)`

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java [hunk 2]
```diff
@@ -278,6 +278,20 @@ public void testInsertIntoNotNullColumn()
         assertUpdate("DROP TABLE test_insert_not_null");
     }
 
+    @Test
+    public void testInsertForDefaultColumn()
+            throws Exception
+    {
+        execute("CREATE TABLE tpch.default_table (a bigint, b bigint default 40)");
```
Comment (kokosing) [inline presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java]: add this test to abstract smoke tests

File: presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java [hunk 3]
```diff
@@ -419,6 +419,11 @@ default ConnectorInsertTableHandle beginInsert(ConnectorSession session, Connect
         throw new PrestoException(NOT_SUPPORTED, "This connector does not support inserts");
     }
 
+    default ConnectorInsertTableHandle beginInsert(ConnectorSession session, ConnectorTableHandle tableHandle, List<ColumnHandle> columnsToBeInserted)
```
Comment (kokosing) [inline presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java]: are we going to support both ways? Which one is prefered? Should we deprecate (or even remove) the other `beginInsert` method?

Comment (Praveen2112) [reply presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java]: We can deprecate the other `beginInsert` method and then remove it

Comment (kokosing) [reply presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java]: Why not to remove it now?

Comment (Praveen2112) [reply presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java]: It might be a backward incompatible change so was planning to remove them after we moved all the connector API to that model. Can we handle it in the same PR or ?

Comment (kokosing) [reply presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java]: No. Let's just deprecate the other method.

**Comment by kokosing** [comment]
> Does Hive connector supports default values?

I mean also Hive in general.

**Comment by Praveen2112** [comment]
I think Hive doesn't support having a default value for its column

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java (hunk 1 above)
Comment (Praveen2112) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java]: Have added it

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java (hunk 2 above)
Comment (Praveen2112) [reply presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java]: but the problem in moving it to abstract test is that not all connectors support adding a default value for its column

File: presto-main/src/main/java/io/prestosql/sql/planner/LogicalPlanner.java [hunk 4]
```diff
@@ -347,8 +346,15 @@ private RelationPlan createInsertPlan(Analysis analysis, Insert insertStatement)
                     assignments.put(output, cast);
                 }
             }
+            visibleColumnMetadataBuilder.add(column);
+            insertColumnsBuilder.add(columns.get(column.getName()));
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/LogicalPlanner.java]: This builder gets the columns from the `INSERT` clause _and_ all the visible columns not mentioned in the `INSERT` clause which don't have a default value. Please add tests where columns without default value are not in the `INSERT` clause.

File: presto-main/src/main/java/io/prestosql/sql/planner/LogicalPlanner.java [hunk 5]
```diff
@@ -331,6 +327,9 @@ private RelationPlan createInsertPlan(Analysis analysis, Insert insertStatement)
             Symbol output = symbolAllocator.newSymbol(column.getName(), column.getType());
             int index = insert.getColumns().indexOf(columns.get(column.getName()));
             if (index < 0) {
+                if (column.hasDefault()) {
+                    continue;
+                }
```
Comment (kasiafi) [inline presto-main/src/main/java/io/prestosql/sql/planner/LogicalPlanner.java]: These columns are missing in the resulting scope and RelationPlan.

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java (hunk 1 above)
Comment (kasiafi) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java]: Please add tests for:
- column without default value (not inserted into)
- insert into other than first column
- other combination (`INSERT INTO default_table (b, a)...`)

File: presto-main/src/main/java/io/prestosql/metadata/Metadata.java [hunk 6]
```diff
@@ -243,7 +243,7 @@
     /**
      * Begin insert query
      */
-    InsertTableHandle beginInsert(Session session, TableHandle tableHandle);
+    InsertTableHandle beginInsert(Session session, TableHandle tableHandle, List<ColumnHandle> columnHandles);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/metadata/Metadata.java]: cmt msg:
> `Skip adding null during Insert if the column has default value`

This commit title is wrong, as this should have nothing to do with default values in JDBC connectors.
Simply, we should INSERT only to columns which user mentioned in their INSERT query.
What happens beyond that (eg defaulting) should happen in the connector or in RDBMS.

Rename to eg

```
Pass information about columns being inserted to connector
```

File: presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java [hunk 7]
```diff
@@ -768,13 +768,13 @@ public OutputTableHandle beginCreateTable(Session session, String catalogName, C
     }
 
     @Override
-    public InsertTableHandle beginInsert(Session session, TableHandle tableHandle)
+    public InsertTableHandle beginInsert(Session session, TableHandle tableHandle, List<ColumnHandle> columnsToBeInserted)
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java]: just call it `columns`

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 8]
```diff
@@ -336,14 +336,6 @@ protected Scope visitInsert(Insert insert, Optional<Scope> scope)
                     .map(ColumnMetadata::getName)
                     .collect(toImmutableList());
 
-            // analyze target table layout
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java]: Why this change?

File: presto-main/src/main/java/io/prestosql/sql/planner/plan/TableWriterNode.java [hunk 9]
```diff
@@ -257,17 +258,24 @@ public String toString()
             extends WriterTarget
     {
         private final TableHandle handle;
+        private final List<ColumnHandle> columnsToBeInserted;
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/plan/TableWriterNode.java]: just `columns`

File: presto-main/src/test/java/io/prestosql/metadata/AbstractMockMetadata.java [hunk 10]
```diff
@@ -305,7 +305,7 @@ public void cleanupQuery(Session session)
     }
 
     @Override
-    public InsertTableHandle beginInsert(Session session, TableHandle tableHandle)
+    public InsertTableHandle beginInsert(Session session, TableHandle tableHandle, List<ColumnHandle> columnsToBeInserted)
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/metadata/AbstractMockMetadata.java]: `columns`

File: presto-spi/src/main/java/io/prestosql/spi/connector/ColumnMetadata.java [hunk 11]
```diff
@@ -30,6 +30,7 @@
     private final String name;
     private final Type type;
     private final boolean nullable;
+    private final boolean hasDefault;
```
Comment (findepi) [inline presto-spi/src/main/java/io/prestosql/spi/connector/ColumnMetadata.java]: This is a wrong abstraction. 

Think about a PostrgeSQL connector and PostrgeSQL table:

```
create table t (
   a bigint,
   b bigint,
   c bigint,
   d bigint default some_postgres_udf());
```

and users run this in Presto:

```
INSERT INTO t(a, b) VALUES (42, 13);
```

Currently Presto creates temporary table to hold the data being inserted.

- the temporary table needs to have `a`, `b` columns to hold the values 
- the temporary table must not have `d` column so that we can leverage PostgreSQL's calculating `some_postgres_udf()`
- the temporary table should not have `c` column, so that we don't write NULLs unnecessarily and the temporary table, inflating the data to be copied at the commit phase.

Thus, it doesn't matter whether column has `DEFAULT` clause or not.
What matters is whether user mentioned the column in their INSERT.

Of course, this is not so simple for Hive connector. In  that case we're responsible for storage ourselves.
We need to fill NULLs for columns not being explicitly INSERTed.
To be backward compatible, we should be able to do this on connector's behalf in the engine.
I don't know yet how exactly this should be addressed.

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java (hunk 1 above)
Comment (findepi) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java]: Please also add column `c` (last) which you insert to.
We need to make sure data is not being "shifted" from `c` to `b`.

Also, please make sure we have tests like that in `TestHiveIntegrationSmokeTest`, as Hive's handling of "extra" columns needs to be different than in any other existing connector.

File: presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java (hunk 2 above)
Comment (findepi) [reply presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java]: I think it's safe to add it in the `AbstractTestIntegrationSmokeTest` and override it for connectors that don't support this.

some day we will have abstract smoke test for JDBC connectors only.
